File: 0835-image-overlap/0835-image-overlap.py
class Solution:
    def largestOverlap(self, img1: List[List[int]], img2: List[List[int]]) -> int:
        list1 = [(r , c) for r in range(len(img1)) for c in range(len(img1)) if img1[r][c] == 1]
        list2 = [(r , c) for r in range(len(img1)) for c in range(len(img1)) if img2[r][c] == 1]

        vectors = []
        for r1 , c1 in list1:
            for r2 , c2 in list2:
                vectors.append((r2 - r1 , c2 - c1))
        if not vectors:
            return 0
        vector_counts = Counter(vectors)
        return max(vector_counts.values())

File: 0835-image-overlap/0835-image-overlap.cpp
class Solution {
public:
    int largestOverlap(vector<vector<int>>& img1, vector<vector<int>>& img2) {
        int n = img1.size();
        vector<pair<int , int>> nonzeros1;
        vector<pair<int , int>> nonzeros2;

        for(int r = 0; r < n; r++){
            for(int c = 0; c < n; c++){
                if(img1[r][c] == 1){
                    nonzeros1.push_back({r , c});
                }
                if(img2[r][c] == 1){
                    nonzeros2.push_back({r , c});
                }
            }
        }

        map<pair<int , int> , int> vectorCounts;
        int maxOverlap = 0;

        for(auto& p1 : nonzeros1){
            for(auto& p2 : nonzeros2){
                pair<int , int> vec = {p2.first - p1.first , p2.second - p1.second};
                vectorCounts[vec]++;
                maxOverlap = max(maxOverlap , vectorCounts[vec]);
            }
        }
        return maxOverlap;
    }
};